pmlc/target/x86/pipeline.cc

```diff
@@ -104,25 +84,124 @@ struct ConvertStandardToLLVMPass
     conversion::stdx_to_llvm::populateStdXToLLVMConversionPatterns(
         typeConverter, patterns);
     populateOpenMPToLLVMConversionPatterns(context, typeConverter, patterns);
-    patterns.insert<ExtractLoweringPattern>(context);
 
     LLVMConversionTarget target(*context);
     target.addDynamicallyLegalOp<omp::ParallelOp>([&](omp::ParallelOp op) {
       return typeConverter.isLegal(&op.getRegion());
     });
     target.addLegalOp<omp::TerminatorOp, omp::TaskyieldOp, omp::FlushOp,
                       omp::BarrierOp, omp::TaskwaitOp>();
-    target.addDynamicallyLegalOp<LLVM::ExtractValueOp>(
-        [](LLVM::ExtractValueOp op) {
-          return !mlir::dyn_cast_or_null<LLVM::InsertValueOp>(
-              op.container().getDefiningOp());
-        });
     if (failed(applyPartialConversion(module, target, patterns))) {
       signalPassFailure();
     }
   }
 };
 
+// OpenMP has issues passing values through to OpenMP blocks.  As a workaround,
+// we have a simple pass to smuggle values that cross the boundary via an
+// alloca()'d struct.
+struct OpenMPWorkaroundPass final
+    : public OpenMPWorkaroundBase<OpenMPWorkaroundPass> {
+  void runOnOperation() final {
+    LLVM::LLVMFuncOp funcOp = getOperation();
+    OpBuilder builder{&getContext()};
+    funcOp.walk([&](omp::ParallelOp parOp) {
+      // The values used to construct the smuggling struct.
+      SmallVector<LLVM::LLVMType, 8> fieldTypes;
+      SmallVector<Value, 8> fieldSrcs;
+
+      // A mapping from original values to the OpOperands within the
+      // parallel region that use those values, for value replacement.
+      std::unordered_map<void *, std::list<OpOperand *>> valueMap;
+
+      // Find values that cross the loop
```

Comment:
Couldn't we use getUsedValuesDefinedAbove?  That would simplify this code quite a bit.  We will almost certainly have to do that to upstream this.

Reply:
We need the `OpOperand`s, but maybe we can use `visitUsedValuesDefinedAbove`?  I'll try it out.

Reply:
Here's a snippet that I've used in multiple places, so it probably belongs in a utility, but seems similar to what you're trying to accomplish:

```
      llvm::SetVector<Value> sinkCandidates;
      getUsedValuesDefinedAbove(band.getLoopBody(), sinkCandidates);

      llvm::SetVector<Operation*> sunkOperations;
      for (Value candidate : sinkCandidates) {
        collectScalarOps(candidate, &sunkOperations);
      }

      BlockAndValueMapping map;
      for (Operation* op : sunkOperations) {
        OpBuilder builder(band.getLoopBody());
        Operation* clonedOp = builder.clone(*op, map);
        // Only replace uses within the current band.
        for (auto result : llvm::enumerate(op->getResults())) {
          auto replacement = clonedOp->getResult(result.index());
          for (auto& use : llvm::make_early_inc_range(result.value().getUses())) {
            if (use.getOwner()->getParentOfType<AffineParallelOp>() == band) {
              use.set(replacement);
            }
          }
        }
      }
    }
```